plugins-bundled/internal/marketplace-app/src/components/InstallControls.tsx

```diff
@@ -0,0 +1,191 @@
+import React, { useState } from 'react';
+import { css } from '@emotion/css';
+import { gt, satisfies } from 'semver';
+
+import { config } from '@grafana/runtime';
+import { Button, HorizontalGroup, Icon, Select, useStyles2 } from '@grafana/ui';
+import { AppEvents, GrafanaTheme2, OrgRole } from '@grafana/data';
+
+import { Metadata, Plugin } from '../types';
+import { hasRole } from '../helpers';
+import API from '../api';
+
+// This isn't exported in the sdk yet
+// @ts-ignore
+import appEvents from 'grafana/app/core/app_events';
+
+interface Props {
+  localPlugin?: Metadata;
+  remotePlugin: Plugin;
+  slug: string;
+}
+
+export const InstallControls = ({ localPlugin, remotePlugin, slug }: Props) => {
+  const [arch, setArch] = useState<string | undefined>();
+  const [loading, setLoading] = useState(false);
+  const [isInstalled, setIsInstalled] = useState(!!localPlugin);
+  const [shouldUpdate, setShouldUpdate] = useState(
+    remotePlugin?.version && localPlugin?.info.version && gt(remotePlugin?.version!, localPlugin?.info.version!)
+  );
+
+  const styles = useStyles2(getStyles);
+  const api = new API();
+
+  const onInstall = (slug: string, version: string) => {
+    setLoading(true);
+    api.installPlugin(slug, version).finally(() => {
+      setLoading(false);
+      setIsInstalled(true);
+      appEvents.emit(AppEvents.alertSuccess, [`Installed ${remotePlugin?.name}`]);
+    });
```

Comment:
Think instead of using these callbacks you could use https://github.com/streamich/react-use/blob/master/docs/useAsyncFn.md to get the state out of the box without a need for the local loading state.

Reply:
I think I'll look into this in a future PR. For now I opted for Hugo's suggestion of using async/await.